File: Cards/Card.ts
/** @noSelfInFile */

import SlideAnimation from "Assets/Animations/SlideAnimation";
import Asset from "Assets/Asset";
import FontWithPosition from "Assets/Fonts/FontWithPosition";
import IconAsset from "Assets/IconAsset";
import GameManager from "GameManager";
import { exhaustiveGuard, isEmpty } from "Helpers";
import { AssetIds, Ranks, Suits, TextIds } from "../Enums";

export interface CardData {
  id: string;
  suit: Suits;
  rank: Ranks;
  power: number;
  value: number;
  isSelected: boolean;
  cost: number;
  name: string;
}

interface EdelConstructionOptions {
  edelName: string;
  edelPower: number;
  edelValue: number;
  edelRankAssetPath: string;
}

export default abstract class Card {
  gameManager: GameManager;
  id: string;
  suit: Suits;
  rank: Ranks;
  private rankAssetPath: string;
  private power: number;
  private value: number;
  isSelected: boolean = false;
  cost: number;
  private name: string;
  private edelName?: string;
  private edelPower?: number;
  private edelValue?: number;
  private edelRankAssetPath?: string;

  constructor(
    gameManager: GameManager,
    suit: Suits,
    rank: Ranks,
    power: number,
    value: number,
    name: string,
    rankAssetPath: string,
    edelConstructionOptions?: EdelConstructionOptions
  ) {
    this.id = `CARD_${suit}_${rank}_${love.math.random(1000)}`;
    this.gameManager = gameManager;
    this.suit = suit;
    this.rank = rank;
    this.power = power;
    this.value = value;
    this.cost = this.getCost();
    this.rankAssetPath = rankAssetPath;
    this.name = name;
    if (edelConstructionOptions) {
      this.edelName = edelConstructionOptions.edelName;
      this.edelPower = edelConstructionOptions.edelPower;
      this.edelValue = edelConstructionOptions.edelValue;
      this.edelRankAssetPath = edelConstructionOptions.edelRankAssetPath;
    }
  }

  isEqual(otherCard: Card): boolean {
    return this.id === otherCard.id;
  }

  getCost(): number {
    let cost = 0;
    cost += this.getBaseCost();
    return cost;
  }

  getBaseCost(): number {
    return this.getPower() * 10 + this.getValue() * 5;
  }

  save(): CardData {
    return {
      id: this.id,
      suit: this.suit,
      rank: this.rank,
      power: this.power,
      value: this.value,
      isSelected: this.isSelected,
      cost: this.cost,
      name: this.name,
    };
  }

  static load(gameManager: GameManager, data: CardData): Card {
    // Use dynamic require to avoid circular dependency
    const CardGenerator = require("./CardGenerator").default;
    const card = CardGenerator.getNewCard(gameManager, data.rank, data.suit);
    card.id = data.id;
    card.isSelected = data.isSelected;
    card.cost = data.cost;
    return card;
  }

  onClick(): void {
    if (this.isSelected) {
      this.onUnselect();
    } else {
      this.onSelect();
    }
  }

  onSelect(): void {
    if (isEmpty(this.gameManager.board)) {
      return;
    }

    this.isSelected = true;

    this.gameManager.board.addPlayerPower(this.getPower());
    this.gameManager.board.addPlayerValue(this.getValue());

    const slideAssets =
      this.gameManager.board?.cardAssets.getCardAssetList(this);

    this.gameManager.animationManager.startAnimation(
      this.id,
      new SlideAnimation(this.gameManager, this.id, 0.15, 0, -20, slideAssets)
    );

    this.gameManager.board?.updatePrimaryButtonStates();
  }

  onUnselect(): void {
    if (isEmpty(this.gameManager.board)) {
      return;
    }

    this.isSelected = false;

    this.gameManager.board.addPlayerPower(-this.getPower());
    this.gameManager.board.addPlayerValue(-this.getValue());

    const slideAssets =
      this.gameManager.board?.cardAssets.getCardAssetList(this);

    this.gameManager.animationManager.startAnimation(
      this.id,
      new SlideAnimation(this.gameManager, this.id, 0.15, 0, 20, slideAssets)
    );

    this.gameManager.board?.updatePrimaryButtonStates();
  }

  onDiscard(): void {
    if (isEmpty(this.gameManager.board)) {
      return;
    }

    this.gameManager.board.addPlayerPower(-this.getPower());
    this.gameManager.board.addPlayerValue(-this.getValue());
  }

  onHover(asset: Asset): void {
    this.gameManager.assetManager.tooltipManager.addTooltip(
      this.getShortCardInfo(5, 10),
      asset
    );
  }

  onUnhover(asset: Asset): void {
    this.gameManager.assetManager.tooltipManager.hideTooltip();
  }

  getShortCardInfo(x: number, y: number): FontWithPosition[] {
    return [
      new FontWithPosition(
        TextIds.TOOLTIP_CARD_NAME,
        x,
        y,
        `${this.getName()} of ${Card.getSuitName(this.suit)}`
      ),
      new FontWithPosition(
        TextIds.TOOLTIP_CARD_POWER,
        x,
        y + 10,
        this.getPower().toString(),
        {
          icon: IconAsset.getPowerIconAsset(
            this.gameManager,
            AssetIds.TOOLTIP_POWER_ICON
          ),
        }
      ),
      new FontWithPosition(
        TextIds.TOOLTIP_CARD_VALUE,
        x,
        y + 20,
        this.getValue().toString(),
        {
          icon: IconAsset.getValueIconAsset(
            this.gameManager,
            AssetIds.TOOLTIP_VALUE_ICON
          ),
        }
      ),
    ];
  }

  static getSuitName(suit: Suits): string {
    switch (suit) {
      case Suits.HEARTS:
        return "Hearts";
      case Suits.ACORNS:
        return "Acorns";
      case Suits.LEAVES:
        return "Leaves";
      case Suits.BELLS:
        return "Bells";
      default:
        exhaustiveGuard(suit);
    }
  }

  get isEdel(): boolean {
    return (
      !this.gameManager.board?.showingEdelView &&
      this.gameManager.board?.edelCard?.suit === this.suit &&
      this.edelName !== this.name
    );
  }

  getPower(): number {
    return this.isEdel && !isEmpty(this.edelPower)
      ? this.edelPower
      : this.power;
  }

  getValue(): number {
    return this.isEdel && !isEmpty(this.edelValue)
      ? this.edelValue
      : this.value;
  }

  getName(): string {
    return this.isEdel && !isEmpty(this.edelName) ? this.edelName : this.name;
  }

  getRankAssetPath(): string {
    return this.isEdel && !isEmpty(this.edelRankAssetPath)
      ? this.edelRankAssetPath
      : this.rankAssetPath;
  }
}

File: Cards/Card.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__New = ____lualib.__TS__New
local __TS__SetDescriptor = ____lualib.__TS__SetDescriptor
local ____exports = {}
local ____SlideAnimation = require("Assets.Animations.SlideAnimation")
local SlideAnimation = ____SlideAnimation.default
local ____FontWithPosition = require("Assets.FontWithPosition")
local FontWithPosition = ____FontWithPosition.default
local ____IconAsset = require("Assets.IconAsset")
local IconAsset = ____IconAsset.default
local ____Helpers = require("Helpers")
local exhaustiveGuard = ____Helpers.exhaustiveGuard
local isEmpty = ____Helpers.isEmpty
local ____Enums = require("Enums")
local AnimationIds = ____Enums.AnimationIds
local AssetIds = ____Enums.AssetIds
local Suits = ____Enums.Suits
local TextIds = ____Enums.TextIds
____exports.default = __TS__Class()
local Card = ____exports.default
Card.name = "Card"
function Card.prototype.____constructor(self, gameManager, suit, rank, power, value, name, rankAssetPath, edelConstructionOptions)
    self.isSelected = false
    self.id = (((suit .. "_") .. rank) .. "_") .. tostring(love.math.random(1000))
    self.gameManager = gameManager
    self.suit = suit
    self.rank = rank
    self.power = power
    self.value = value
    self.cost = self:getCost()
    self.rankAssetPath = rankAssetPath
    self.name = name
    if edelConstructionOptions then
        self.edelName = edelConstructionOptions.edelName
        self.edelPower = edelConstructionOptions.edelPower
        self.edelValue = edelConstructionOptions.edelValue
        self.edelRankAssetPath = edelConstructionOptions.edelRankAssetPath
    end
end
function Card.prototype.isEqual(self, otherCard)
    return self.id == otherCard.id
end
function Card.prototype.getCost(self)
    local cost = 0
    cost = cost + self:getBaseCost()
    return cost
end
function Card.prototype.getBaseCost(self)
    return self:getPower() * 10 + self:getValue() * 5
end
function Card.prototype.save(self)
    return {
        id = self.id,
        suit = self.suit,
        rank = self.rank,
        power = self.power,
        value = self.value,
        isSelected = self.isSelected,
        cost = self.cost,
        name = self.name
    }
end
function Card.load(self, gameManager, data)
    local CardGenerator = require("Cards.CardGenerator").default
    local card = CardGenerator:getNewCard(gameManager, data.rank, data.suit)
    card.id = data.id
    card.isSelected = data.isSelected
    card.cost = data.cost
    return card
end
function Card.prototype.onClick(self)
    if self.isSelected then
        self:onUnselect()
    else
        self:onSelect()
    end
end
function Card.prototype.onSelect(self)
    if isEmpty(self.gameManager.board) then
        return
    end
    self.isSelected = true
    self.gameManager.board:addPlayerPower(self:getPower())
    self.gameManager.board:addPlayerValue(self:getValue())
    local ____opt_0 = self.gameManager.board
    local slideAssets = ____opt_0 and ____opt_0.cardAssets:getCardAssetList(self)
    self.gameManager.animationManager.animations:set(
        AnimationIds.CARD_SELECT .. self.id,
        __TS__New(
            SlideAnimation,
            0.15,
            0,
            -20,
            slideAssets
        )
    )
    local ____opt_2 = self.gameManager.board
    if ____opt_2 ~= nil then
        ____opt_2:updatePrimaryButtonStates()
    end
end
function Card.prototype.onUnselect(self)
    if isEmpty(self.gameManager.board) then
        return
    end
    self.isSelected = false
    self.gameManager.board:addPlayerPower(-self:getPower())
    self.gameManager.board:addPlayerValue(-self:getValue())
    local ____opt_4 = self.gameManager.board
    local slideAssets = ____opt_4 and ____opt_4.cardAssets:getCardAssetList(self)
    self.gameManager.animationManager.animations:set(
        AnimationIds.CARD_SELECT .. self.id,
        __TS__New(
            SlideAnimation,
            0.15,
            0,
            20,
            slideAssets
        )
    )
    local ____opt_6 = self.gameManager.board
    if ____opt_6 ~= nil then
        ____opt_6:updatePrimaryButtonStates()
    end
end
function Card.prototype.onHover(self, asset)
    self.gameManager.assetManager.tooltipManager:addTooltip(
        {
            __TS__New(
                FontWithPosition,
                TextIds.TOOLTIP_CARD_NAME,
                5,
                10,
                (self:getName() .. " of ") .. ____exports.default:getSuitName(self.suit)
            ),
            __TS__New(
                FontWithPosition,
                TextIds.TOOLTIP_CARD_POWER,
                5,
                20,
                tostring(self:getPower()),
                {icon = IconAsset:getPowerIconAsset(self.gameManager, AssetIds.TOOLTIP_POWER_ICON)}
            ),
            __TS__New(
                FontWithPosition,
                TextIds.TOOLTIP_CARD_VALUE,
                5,
                30,
                tostring(self:getValue()),
                {icon = IconAsset:getValueIconAsset(self.gameManager, AssetIds.TOOLTIP_VALUE_ICON)}
            )
        },
        asset
    )
end
function Card.prototype.onUnhover(self, asset)
    self.gameManager.assetManager.tooltipManager:hideTooltip()
end
function Card.getSuitName(self, suit)
    repeat
        local ____switch19 = suit
        local ____cond19 = ____switch19 == Suits.HEARTS
        if ____cond19 then
            return "Hearts"
        end
        ____cond19 = ____cond19 or ____switch19 == Suits.ACORNS
        if ____cond19 then
            return "Acorns"
        end
        ____cond19 = ____cond19 or ____switch19 == Suits.LEAVES
        if ____cond19 then
            return "Leaves"
        end
        ____cond19 = ____cond19 or ____switch19 == Suits.BELLS
        if ____cond19 then
            return "Bells"
        end
        do
            exhaustiveGuard(suit)
        end
    until true
end
function Card.prototype.getPower(self)
    return self.isEdel and not isEmpty(self.edelPower) and self.edelPower or self.power
end
function Card.prototype.getValue(self)
    return self.isEdel and not isEmpty(self.edelValue) and self.edelValue or self.value
end
function Card.prototype.getName(self)
    return self.isEdel and not isEmpty(self.edelName) and self.edelName or self.name
end
function Card.prototype.getRankAssetPath(self)
    return self.isEdel and not isEmpty(self.edelRankAssetPath) and self.edelRankAssetPath or self.rankAssetPath
end
__TS__SetDescriptor(
    Card.prototype,
    "isEdel",
    {get = function(self)
        local ____opt_8 = self.gameManager.board
        local ____temp_14 = not (____opt_8 and ____opt_8.showingEdelView)
        if ____temp_14 then
            local ____opt_12 = self.gameManager.board
            local ____opt_10 = ____opt_12 and ____opt_12.edelCard
            ____temp_14 = (____opt_10 and ____opt_10.suit) == self.suit
        end
        return ____temp_14 and self.edelName ~= self.name
    end},
    true
)
return ____exports
